### PR TITLE
Make base64_decode relaxed mode less relaxed

### PR DESCRIPTION
**Pull request by karel-m**
### Checklist

* [x] documentation is added or updated
* [x] tests are added or updated

### CODE REVIEW

**Comment by karel-m** [comment]
> @sjaeckel wrote off-list:
> I think we should try to keep this super relaxed mode if possible, as that's the one that is compatible with the old 1.17 behavior, right?

IMO the current super relaxed mode may be dangerous when you are decoding data from external sources (e.g. from a user request) because we are "able" to decode nearly any invalid input data. I am open to discussion what other characters should be ignored in relaxed mode.

**Comment by sjaeckel** [comment]
> IMO the current super relaxed mode may be dangerous when you are decoding data from external sources (e.g. from a user request)...

I fully agree! ... and that's also what is said in [RFC 4648](https://tools.ietf.org/html/rfc4648#section-12) ...

But I also think that we shouldn't break existing code or at least give the possibility to enable the old behavior if required... so we've several options I guess...

1. add a new API function `base64_insane_decode()` and fix the original
1. add a new API function `base64_sane_decode()` and keep the original
1. add a way to set the level of sanity checking... 
a. globally via a global 
b. globally via `base64_set_decode_sanity()` 
c. on the API level via exporting the sanity parameter
d. at compile time
1. or a combination of some of the above or something else I didn't think of?

**Comment by karel-m** [comment]
From the options above I prefer `base64_sane_decode()` approach

**Comment by sjaeckel** [comment]
> From the options above I prefer `base64_sane_decode()` approach

:+1: 

`base64_sane_decode()` shall it be

so we have now 3 possible values `enum(insane, relaxed, strict)` where the default `base64_decode()` is `insane` :D

**Comment by karel-m** [comment]
`base64_sane_decode` or `base64_relaxed_decode`?

**Comment by sjaeckel** [comment]
I think `base64_sane_decode()` is more appropriate as `base64_relaxed_decode()` is basically lying because `base64_decode()` is more relaxed...

File: src/misc/base64/base64_decode.c [hunk 1]
```diff
@@ -89,20 +95,29 @@ static int _base64_decode_internal(const char *in,  unsigned long inlen,
 
    g = 0; /* '=' counter */
    for (x = y = z = t = 0; x < inlen; x++) {
+       if (in[x] == 0 && x == (inlen - 1) && mode != strict) {
```
Comment (sjaeckel) [inline src/misc/base64/base64_decode.c]: can you please add parentheses?

File: src/misc/base64/base64_decode.c [hunk 2]
```diff
@@ -118,7 +133,7 @@ static int _base64_decode_internal(const char *in,  unsigned long inlen,
 
    if (y != 0) {
       if (y == 1) return CRYPT_INVALID_PACKET;
-      if ((y + g) != 4 && is_strict && map != map_base64url) return CRYPT_INVALID_PACKET;
+      if ((y + g) != 4 && (mode == strict) && map != map_base64url) return CRYPT_INVALID_PACKET;
```
Comment (sjaeckel) [inline src/misc/base64/base64_decode.c]: here as well please

File: src/misc/base64/base64_decode.c [hunk 3]
```diff
@@ -156,6 +171,20 @@ int base64_strict_decode(const char *in,  unsigned long inlen,
 {
    return _base64_decode_internal(in, inlen, out, outlen, map_base64, strict);
 }
+
+/**
+   Relaxed base64 decode a block of memory
```
Comment (sjaeckel) [inline src/misc/base64/base64_decode.c]: Probably this (and the URL variant) comment should be changed?

File: tests/base64_test.c [hunk 4]
```diff
@@ -47,39 +49,68 @@ int base64_test(void)
...
-         {"vuiSPKIl8PiR5O-rC4z9_xTQKZ0=", 1},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0==", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0===", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0====", 0},
-         {"vuiS*=PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
-         {"vuiS*==PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
-         {"vuiS*===PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
+         {"vuiSPKIl8PiR5O-rC4z9_xTQKZ0", strict},                       /* 0 */
```
Comment (sjaeckel) [inline tests/base64_test.c]: :+1:

File: tests/base64_test.c [hunk 5]
```diff
@@ -47,39 +49,68 @@ int base64_test(void)
...
           DO(base64url_strict_decode(url_cases[x].s, slen1, tmp, &l1));
-       else
+          DO(do_compare_testvector(tmp, l1, special_case, sizeof(special_case) - 1, "base64url_strict_decode", x));
+          DO(base64url_sane_decode(url_cases[x].s, slen1, tmp, &l1));
+          DO(do_compare_testvector(tmp, l1, special_case, sizeof(special_case) - 1, "base64url_sane_decode/strict", x));
+          DO(base64url_decode(url_cases[x].s, slen1, tmp, &l1));
+          DO(do_compare_testvector(tmp, l1, special_case, sizeof(special_case) - 1, "base64url_decode/strict", x));
+       }
+       else if(url_cases[x].flag == relaxed) {
+          DO(base64url_strict_decode(url_cases[x].s, slen1, tmp, &l1) == CRYPT_INVALID_PACKET ? CRYPT_OK : CRYPT_FAIL_TESTVECTOR);
```
Comment (sjaeckel) [inline tests/base64_test.c]: I already thought about it while writing the last base16 testcase... should we add this?

```
diff --git a/tests/common.h b/tests/common.h
index 1ba8820..f63e26b 100644
--- a/tests/common.h
+++ b/tests/common.h
@@ -18,2 +18,3 @@ extern prng_state yarrow_prng;
 #define DOX(x, str) do { fprintf(stderr, "%s - %s:\n", #x, (str)); run_cmd((x), __LINE__, __FILE__, #x, (str)); } while (0)
+#define SHOULD_FAIL(x) do { fprintf(stderr, "%s:\n", #x); run_cmd((x) != CRYPT_OK ? CRYPT_OK : CRYPT_FAIL_TESTVECTOR, __LINE__, __FILE__, #x, NULL); } while (0)
 #else
@@ -21,2 +22,3 @@ extern prng_state yarrow_prng;
 #define DOX(x, str) do { run_cmd((x), __LINE__, __FILE__, #x, (str)); } while (0)
+#define SHOULD_FAIL(x) do { run_cmd((x) != CRYPT_OK ? CRYPT_OK : CRYPT_FAIL_TESTVECTOR, __LINE__, __FILE__, #x, NULL); } while (0)
 #endif
```

File: doc/crypt.tex [hunk 6]
```diff
@@ -6579,6 +6590,9 @@ \subsection{URL--safe 'base64url' encoding}
 
 int base64url_strict_decode(   const char *in,  unsigned long len,
                             unsigned char *out, unsigned long *outlen);
+
+int base64url_sane_decode(   const char *in,  unsigned long len,
```
Comment (sjaeckel) [inline doc/crypt.tex]: should we add the URL variants to the index?

File: tests/base64_test.c (hunk 5 above)
Comment (karel-m) [reply tests/base64_test.c]: Maybe, but perhaps as `SHOULD_FAIL(x, expected_retval)`

File: src/misc/base64/base64_decode.c (hunk 3 above)
Comment (karel-m) [reply src/misc/base64/base64_decode.c]: done

File: doc/crypt.tex (hunk 6 above)
Comment (karel-m) [reply doc/crypt.tex]: done

File: tests/base64_test.c (hunk 5 above)
Comment (sjaeckel) [reply tests/base64_test.c]: I also thought about `EXPECT(x, e)` but in my opinion that's another use-case.
Most cases where we do negative tests shouldn't care about the error-code returned, it should just be ensured that the result is not OK.

File: src/misc/base64/base64_decode.c [hunk 7]
```diff
@@ -130,7 +145,7 @@ static int _base64_decode_internal(const char *in,  unsigned long inlen,
 
 #if defined(LTC_BASE64)
 /**
-   Relaxed base64 decode a block of memory
+   Dangerously relaxed base64 decode a block of memory
```
Comment (sjaeckel) [inline src/misc/base64/base64_decode.c]: :) well put!

**Comment by karel-m** [comment]
Is it ok?

**Pull request merged**